File: src/auth_helpers.py
"""Shared auth helpers used by all route files."""

import os
from dataclasses import dataclass
from typing import Optional
from fastapi import Request, HTTPException

from src.owner_identity import (
    auth_disabled,
    effective_storage_owner,
    is_request_sentinel_owner,
)


@dataclass(frozen=True)
class RequestCapability:
    """Immutable request authority passed through chat execution helpers.

    A bearer token that has the narrow ``chat`` scope is still a pure chat
    capability.  It may complete the synchronous model call, but it cannot
    create detached execution, emit interactive events, or schedule follow-up
    work that would run after the request's authorization context is gone.
    Cookie and AUTH_ENABLED=false requests retain the existing interactive
    behavior.
    """

    principal: str
    owner: Optional[str]
    is_bearer: bool
    allow_deferred_work: bool
    allow_detached_execution: bool
    allow_message_events: bool
    allow_auto_naming: bool


def is_bearer_principal(request: Request) -> bool:
    """Return whether the request is attributable to an API-token principal.

    The auth middleware stamps ``state.api_token`` for a verified token.  The
    header/sentinel checks keep direct endpoint calls and auth-disabled
    alternate entry points fail-closed instead of treating the ``api``
    sentinel as a normal cookie user.
    """
    state = getattr(request, "state", None)
    if getattr(state, "api_token", False) is True:
        return True
    current_user = getattr(state, "current_user", None)
    if isinstance(current_user, str) and current_user.strip().casefold() == "api":
        return True
    try:
        auth_header = request.headers.get("authorization", "")
    except Exception:
        auth_header = ""
    return isinstance(auth_header, str) and auth_header.strip().casefold().startswith("bearer ody_")


def get_current_user(request: Request) -> Optional[str]:
    """Get current username from request state (set by auth middleware)."""
    state = getattr(request, "state", None)
    return getattr(state, "current_user", None)


def effective_user(request: Request) -> Optional[str]:
    """The real human behind the request, for ownership/attribution.

    Cookie sessions resolve to the logged-in username. Bearer ``ody_`` callers
    come through as the sandboxed pseudo-user "api" so they can't wander into
    cookie/user routes by default, but their token was minted by, and belongs
    to, a real owner stamped on ``request.state.api_token_owner``. Routes that
    should attribute a token's actions to that owner (sessions, chat history)
    call this instead of :func:`get_current_user`, so a paired client sees and
    creates the SAME data as the owner's desktop UI rather than a separate
    "api"-owned silo.

    For cookie sessions this is identical to :func:`get_current_user`, so
    swapping a route over is a no-op for browser users. A bearer token with no
    owner falls back to :func:`get_current_user` (the "api" pseudo-user), so it
    never escalates.
    """
    if _is_api_token_request(request):
        state = getattr(request, "state", None)
        owner = getattr(state, "api_token_owner", None)
        if isinstance(owner, str) and owner.strip():
            return owner.strip()
    return get_current_user(request)


def _is_api_token_request(request: Request) -> bool:
    """Return True when the request has a bearer API-token principal."""
    return is_bearer_principal(request)


def request_capability(request: Request) -> RequestCapability:
    """Build the one request capability shared by chat downstream helpers."""
    bearer = is_bearer_principal(request)
    return RequestCapability(
        principal="bearer" if bearer else "interactive",
        owner=effective_user(request),
        is_bearer=bearer,
        allow_deferred_work=not bearer,
        allow_detached_execution=not bearer,
        allow_message_events=not bearer,
        allow_auto_naming=not bearer,
    )


def require_api_token_owner(request: Request) -> str:
    """Return a real owner for a bearer request, failing closed otherwise.

    The middleware normally resolves token owners against configured human
    accounts. Keep that invariant at route boundaries too: direct endpoint
    tests, alternate ASGI entry points, and future middleware changes must not
    turn a request sentinel or an ownerless token into a durable/executable
    owner.
    """
    state = getattr(request, "state", None)
    owner = getattr(state, "api_token_owner", None)
    if (
        not isinstance(owner, str)
        or not owner.strip()
        or is_request_sentinel_owner(owner)
    ):
        raise HTTPException(403, "API token has no owner")
    normalized_owner = owner.strip()
    # The normal auth middleware has already resolved this identity from the
    # token row. Keep the same invariant for direct endpoint calls and
    # alternate ASGI entry points when a configured auth manager is available.
    auth_state = getattr(getattr(request, "app", None), "state", None)
    auth_manager = getattr(auth_state, "auth_manager", None)
    users = getattr(auth_manager, "users", None)
    if (
        getattr(auth_manager, "is_configured", False)
        and isinstance(users, dict)
        and normalized_owner.casefold() not in {
            str(username).strip().casefold() for username in users
        }
    ):
        raise HTTPException(403, "API token owner is not a configured user")
    return normalized_owner


def require_api_token_scope(request: Request, required_scope: str) -> Optional[str]:
    """Require one declared scope for bearer callers; leave browser callers unchanged."""
    if not _is_api_token_request(request):
        return effective_user(request)
    state = getattr(request, "state", None)
    raw_scopes = getattr(state, "api_token_scopes", None)
    if isinstance(raw_scopes, (list, tuple, set, frozenset)):
        scopes = {
            value.strip().casefold()
            for value in raw_scopes
            if isinstance(value, str) and value.strip()
        }
    else:
        scopes = set()
    normalized_scope = str(required_scope or "").strip().casefold()
    if not normalized_scope or normalized_scope not in scopes:
        raise HTTPException(403, f"API token missing required scope: {required_scope}")
    return require_api_token_owner(request)


def require_chat_scope(request: Request) -> Optional[str]:
    """FastAPI dependency for owner-scoped chat/session routes."""
    return require_api_token_scope(request, "chat")


def require_interactive_request(request: Request) -> Optional[str]:
    """Reject bearer integrations from browser-only agent/control surfaces.

    This is deliberately a bearer-principal gate rather than an authentication
    requirement. Cookie sessions and AUTH_ENABLED=false keep their existing
    route behavior, while API tokens cannot enter routes that start, resume,
    approve, or otherwise control interactive agent work.
    """
    current_user = get_current_user(request)
    if is_bearer_principal(request):
        raise HTTPException(403, "API tokens cannot use this interactive surface")
    return current_user


def require_non_bearer_request(request: Request) -> Optional[str]:
    """Reject bearer principals while preserving cookie/local route behavior."""
    if is_bearer_principal(request):
        raise HTTPException(403, "API tokens cannot use this host-control surface")
    return get_current_user(request)


def enforce_api_token_chat_controls(
    request: Request,
    *,
    mode: str,
    plan_mode: bool,
    approval_id: object,
    allow_bash: object,
) -> bool:
    """Reject bearer-token controls that can enter or authorize agent execution."""
    is_api_token = _is_api_token_request(request)
    if is_api_token and (
        approval_id
        or plan_mode
        or mode != "chat"
        or str(allow_bash or "").lower() == "true"
    ):
        raise HTTPException(403, "API tokens cannot use agent tools or approve tool calls")
    return is_api_token


def require_authenticated_request(request: Request) -> str:
    """Allow either a browser session or a valid bearer API token.

    This is intentionally narrower than :func:`require_user`: use it only for
    routes that need authentication but do not read or mutate owner-scoped
    user data. Owner-scoped routes should use ``require_user`` for browser
    sessions or their own API-token scope/owner gate.
    """
    if is_bearer_principal(request):
        return require_api_token_owner(request)
    return require_user(request)


def _auth_disabled() -> bool:
    """True when the operator has explicitly turned off auth via .env.
    Mirrors the AUTH_ENABLED parse in app.py / core/middleware.py so the
    three call sites agree on what "off" means."""
    return auth_disabled()


def storage_owner_for_request(request: Request) -> Optional[str]:
    """Resolve the storage owner for code paths that need an owner bucket.

    This does not replace route authentication. It only gives auth-disabled
    no-login mode a stable storage identity instead of writing new data as
    legacy NULL/ownerless state.
    """
    return effective_storage_owner(effective_user(request))


def require_user(request: Request) -> str:
    """FastAPI dependency: reject unauthenticated callers when the upstream
    auth middleware was bypassed unexpectedly (e.g. SSRF from a sibling
    service). Returns the resolved username, or "" in single-user / anonymous
    modes where no username is available.

    The three "" cases are:
      1. AUTH_ENABLED=false — the operator explicitly turned auth off.
         The full /login flow is skipped (issue #622), so route-level
         require_user must let the request through too instead of 401-ing
         and forcing the browser to /login.
      2. Unconfigured first-run + loopback caller — pre-setup access from
         localhost so the operator can hit the SPA before creating the
         first admin.
      3. LOCALHOST_BYPASS=true + loopback caller — documented dev bypass.

    Use this on routes that touch user data so middleware misconfig can't
    open them up.
    """
    if is_bearer_principal(request):
        raise HTTPException(403, "API tokens must use a scope-aware API route")

    u = get_current_user(request)
    if u:
        return u
    # Operator-disabled auth: honor it at the route layer too. Without this,
    # routes that depend on require_user 401, the front-end fetch wrapper
    # redirects to /login, and the user sees a login page despite
    # AUTH_ENABLED=false (issue #622). Docker / reverse-proxy deployments
    # hit this because requests arrive from a non-loopback client.host, so
    # the loopback fall-through below never fires.
    if _auth_disabled():
        return ""
    auth_mgr = getattr(request.app.state, "auth_manager", None)
    client = getattr(request, "client", None)
    host = (client.host if client else "") or ""
    is_loopback = host in ("127.0.0.1", "::1", "localhost")
    # LOCALHOST_BYPASS=true is the dev-only "I'm on loopback, skip auth"
    # switch. Mirror the middleware so routes don't 401 the same caller
    # the middleware just let through.
    if is_loopback and os.getenv("LOCALHOST_BYPASS", "false").lower() == "true":
        return ""
    if auth_mgr is not None and getattr(auth_mgr, "is_configured", False):
        raise HTTPException(401, "Not authenticated")
    # Unconfigured / first-run mode: only allow loopback callers.
    if is_loopback:
        return ""
    raise HTTPException(401, "Not authenticated")


def require_privilege(request: Request, key: str) -> str:
    """Reject callers whose `auth.json` privilege flag for `key` is False.
    Returns the username so the route handler can keep using it.

    Admins always have every privilege via `auth_manager.get_privileges`
    (which returns ADMIN_PRIVILEGES wholesale), so this is a no-op for
    them. In unauthenticated single-user mode (`require_user` returns ""),
    privileges aren't enforced.
    """
    user = require_user(request)
    if not user:
        return user
    auth_mgr = getattr(request.app.state, "auth_manager", None)
    if auth_mgr is None:
        return user
    try:
        privs = auth_mgr.get_privileges(user) or {}
    except Exception:
        return user
    if not isinstance(privs, dict):
        privs = {}
    # True = permitted; missing key defaults to permitted (unknown privileges
    # fail open — the UI gates display-side).
    if not privs.get(key, True):
        raise HTTPException(403, f"Your account is not allowed to {key.replace('_', ' ')}.")
    return user


def owner_filter(query, model_cls, user: str, *, include_shared: bool = True):
    """Filter `query` so only rows owned by `user` (and optionally null-owner
    'shared' rows) come through. No-op when `user` is empty (single-user
    mode). Returns the modified query."""
    if not user:
        return query
    if include_shared:
        return query.filter((model_cls.owner == user) | (model_cls.owner == None))  # noqa: E711
    return query.filter(model_cls.owner == user)

File: rust/src/src/auth_helpers.rs
// src/auth_helpers.rs  <- src/auth_helpers.py
//! Shared auth helpers used by all route files.

use crate::core::auth::AuthManager;
use crate::error::{PyError, PyResult};
use crate::pyos as os;

/// A faithful stand-in for the FastAPI `Request` object the helpers receive.
///
/// The Python reads three things off the request via `getattr`:
///   * `request.state.current_user`     — set by the auth middleware,
///   * `request.app.state.auth_manager` — the shared [`AuthManager`],
///   * `request.client.host`            — the peer address.
/// We model exactly that surface so the logic below reads line-by-line.
pub struct Request<'a> {
    /// `request.state.current_user` (`None` when middleware didn't set it).
    pub current_user: Option<String>,
    /// `request.app.state.auth_manager` (`None` when not wired up).
    pub auth_manager: Option<&'a AuthManager>,
    /// `request.client.host` (`None` when there is no client, e.g. test calls).
    pub client_host: Option<String>,
}

/// Get current username from request state (set by auth middleware).
pub fn get_current_user(request: &Request) -> Option<String> {
    // getattr(request.state, 'current_user', None)
    request.current_user.clone()
}

/// The real human behind the request, for ownership/attribution.
///
/// Cookie sessions resolve to the logged-in username. Bearer `ody_` callers
/// come through as the sandboxed pseudo-user `"api"` so they can't wander into
/// cookie/user routes by default, but their token was minted by, and belongs
/// to, a real owner stamped on `request.state.api_token_owner`. Routes that
/// should attribute a token's actions to that owner (sessions, chat history)
/// call this instead of [`get_current_user`], so a paired client sees and
/// creates the SAME data as the owner's desktop UI rather than a separate
/// `"api"`-owned silo.
///
/// For cookie sessions this is identical to [`get_current_user`], so swapping a
/// route over is a no-op for browser users. A bearer token with no owner falls
/// back to [`get_current_user`] (the `"api"` pseudo-user), so it never escalates.
///
/// `api_token` / `api_token_owner` are the two facts the auth gate stamps on
/// `request.state` (in the axum port, the `web::ApiToken` extension's `present`
/// / `owner` fields); they are threaded in explicitly rather than carried on the
/// [`Request`] stand-in so the existing `require_user` / `require_privilege`
/// call sites keep their three-field `Request` literal unchanged.
//
// def effective_user(request):
//     if getattr(request.state, "api_token", False):
//         owner = getattr(request.state, "api_token_owner", None)
//         if owner:
//             return owner
//     return get_current_user(request)
pub fn effective_user(
    request: &Request,
    api_token: bool,
    api_token_owner: Option<&str>,
) -> Option<String> {
    // if getattr(request.state, "api_token", False):
    if api_token {
        // owner = getattr(request.state, "api_token_owner", None)
        // if owner: return owner  — `if owner` is falsy for both None and ""
        if let Some(owner) = api_token_owner {
            if !owner.is_empty() {
                return Some(owner.to_string());
            }
        }
    }
    // return get_current_user(request)
    get_current_user(request)
}

/// True when the operator has explicitly turned off auth via `.env`.
/// Mirrors the `AUTH_ENABLED` parse in `app.py` / `core/middleware.py` so the
/// three call sites agree on what "off" means.
//
// def _auth_disabled() -> bool:
//     return os.getenv("AUTH_ENABLED", "true").lower() == "false"
fn auth_disabled() -> bool {
    os::getenv("AUTH_ENABLED", "true").to_lowercase() == "false"
}

/// FastAPI dependency: reject unauthenticated callers when the upstream auth
/// middleware was bypassed unexpectedly (e.g. SSRF from a sibling service).
/// Returns the resolved username, or `""` in single-user / anonymous modes
/// where no username is available.
///
/// The three `""` cases are:
///   1. `AUTH_ENABLED=false` — the operator explicitly turned auth off. The
///      full `/login` flow is skipped (issue #622), so route-level
///      `require_user` must let the request through too instead of 401-ing and
///      forcing the browser to `/login`.
///   2. Unconfigured first-run + loopback caller — pre-setup access from
///      localhost so the operator can hit the SPA before creating the first
///      admin.
///   3. `LOCALHOST_BYPASS=true` + loopback caller — documented dev bypass.
///
/// Use this on routes that touch user data so middleware misconfig can't
/// open them up.
pub fn require_user(request: &Request) -> PyResult<String> {
    let u = get_current_user(request);
    if let Some(u) = u {
        if !u.is_empty() {
            return Ok(u);
        }
    }
    // Operator-disabled auth: honor it at the route layer too. Without this,
    // routes that depend on require_user 401, the front-end fetch wrapper
    // redirects to /login, and the user sees a login page despite
    // AUTH_ENABLED=false (issue #622). Docker / reverse-proxy deployments
    // hit this because requests arrive from a non-loopback client.host, so
    // the loopback fall-through below never fires.
    if auth_disabled() {
        return Ok(String::new());
    }
    let auth_mgr = request.auth_manager; // getattr(request.app.state, "auth_manager", None)
    let client = request.client_host.as_deref(); // getattr(request, "client", None)
    let host = client.unwrap_or(""); // (client.host if client else "") or ""
    let is_loopback = host == "127.0.0.1" || host == "::1" || host == "localhost";
    // LOCALHOST_BYPASS=true is the dev-only "I'm on loopback, skip auth"
    // switch. Mirror the middleware so routes don't 401 the same caller
    // the middleware just let through.
    if is_loopback && os::getenv("LOCALHOST_BYPASS", "false").to_lowercase() == "true" {
        return Ok(String::new());
    }
    if let Some(auth_mgr) = auth_mgr {
        if auth_mgr.is_configured() {
            // raise HTTPException(401, "Not authenticated")
            return Err(PyError::other("Not authenticated"));
        }
    }
    // Unconfigured / first-run mode: only allow loopback callers.
    if is_loopback {
        return Ok(String::new());
    }
    // raise HTTPException(401, "Not authenticated")
    Err(PyError::other("Not authenticated"))
}

/// Reject callers whose `auth.json` privilege flag for `key` is False.
/// Returns the username so the route handler can keep using it.
///
/// Admins always have every privilege via `auth_manager.get_privileges`
/// (which returns ADMIN_PRIVILEGES wholesale), so this is a no-op for
/// them. In unauthenticated single-user mode (`require_user` returns ""),
/// privileges aren't enforced.
pub fn require_privilege(request: &Request, key: &str) -> PyResult<String> {
    let user = require_user(request)?;
    if user.is_empty() {
        return Ok(user);
    }
    let auth_mgr = request.auth_manager; // getattr(request.app.state, "auth_manager", None)
    let auth_mgr = match auth_mgr {
        None => return Ok(user),
        Some(m) => m,
    };
    // try: privs = auth_mgr.get_privileges(user) or {}  / except Exception: return user
    // `get_privileges` is infallible in the Rust port, so the `except` arm is
    // unreachable; the merged map is never empty (defaults), so `or {}` is moot.
    // The upstream `if not isinstance(privs, dict): privs = {}` defensive cast is
    // also a no-op here: `get_privileges` returns `Map<String, Value>`, so `privs`
    // is statically always a dict and can never need the `{}` fallback.
    let privs = auth_mgr.get_privileges(&user);
    // True = permitted; missing key defaults to permitted (unknown privileges
    // fail open — the UI gates display-side).
    let permitted = privs.get(key).and_then(|v| v.as_bool()).unwrap_or(true);
    if !permitted {
        // raise HTTPException(403, f"Your account is not allowed to {key.replace('_', ' ')}.")
        return Err(PyError::permission(format!(
            "Your account is not allowed to {}.",
            key.replace('_', " ")
        )));
    }
    Ok(user)
}

/// Filter `query` so only rows owned by `user` (and optionally null-owner
/// 'shared' rows) come through. No-op when `user` is empty (single-user
/// mode). Returns the modified query.
///
// TODO(db): `query` is a SQLAlchemy `Query` and `model_cls` a mapped model;
// the `.filter(...)` chain with `model_cls.owner == user` / `== None` has no
// faithful analogue without pulling in a query builder. We mirror the control
// flow and leave the actual filtering to the DB layer once it exists.
pub fn owner_filter<Q>(query: Q, _model_cls: (), user: &str, include_shared: bool) -> Q {
    if user.is_empty() {
        return query;
    }
    if include_shared {
        // return query.filter((model_cls.owner == user) | (model_cls.owner == None))
        unimplemented!("owner_filter: SQLAlchemy query.filter (shared) — TODO(db)")
    }
    // return query.filter(model_cls.owner == user)
    unimplemented!("owner_filter: SQLAlchemy query.filter — TODO(db)")
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::core::auth::AuthManager;

    /// A request stand-in with no resolved user and the given peer host.
    fn anon_req<'a>(auth: &'a AuthManager, host: Option<&str>) -> Request<'a> {
        Request {
            current_user: None,
            auth_manager: Some(auth),
            client_host: host.map(str::to_string),
        }
    }

    /// `effective_user` is a pure read of the threaded api-token facts + the
    /// stamped `current_user`; it touches no env, so it is safe to run in
    /// parallel with everything else.
    #[test]
    fn effective_user_prefers_token_owner_then_falls_back() {
        let auth = AuthManager::new();

        // Cookie session (not a token): identical to get_current_user.
        let cookie = Request {
            current_user: Some("alice".to_string()),
            auth_manager: Some(&auth),
            client_host: None,
        };
        assert_eq!(effective_user(&cookie, false, None), Some("alice".to_string()));
        // A token-owner on a non-token request is ignored (api_token=false).
        assert_eq!(
            effective_user(&cookie, false, Some("bob")),
            Some("alice".to_string())
        );

        // Bearer ody_ caller stamped as the "api" pseudo-user with a real owner:
        // attributed to the owner, not "api".
        let bearer = Request {
            current_user: Some("api".to_string()),
            auth_manager: Some(&auth),
            client_host: None,
        };
        assert_eq!(
            effective_user(&bearer, true, Some("bob")),
            Some("bob".to_string())
        );
        // Token with no owner (None) or an empty owner: never escalates — falls
        // back to get_current_user (the "api" pseudo-user). `if owner:` is falsy
        // for both None and "".
        assert_eq!(effective_user(&bearer, true, None), Some("api".to_string()));
        assert_eq!(effective_user(&bearer, true, Some("")), Some("api".to_string()));
    }

    /// `require_user`'s env-sensitive branches (AUTH_ENABLED=false, the
    /// LOCALHOST_BYPASS + loopback bypass) plus the unconfigured-loopback /
    /// non-loopback fall-throughs. Env vars are process-global, so all of these
    /// assertions live in ONE test (run serially within this thread) and the
    /// prior values are saved and restored around each mutation.
    #[test]
    fn require_user_env_modes() {
        let auth = AuthManager::new();

        let prev_auth = std::env::var("AUTH_ENABLED").ok();
        let prev_bypass = std::env::var("LOCALHOST_BYPASS").ok();

        let restore = |key: &str, prev: &Option<String>| match prev {
            Some(v) => std::env::set_var(key, v),
            None => std::env::remove_var(key),
        };

        // --- 1. AUTH_ENABLED=false: "" for any caller (even non-loopback),
        // short-circuiting before any is_configured check (issue #622). ---
        std::env::set_var("AUTH_ENABLED", "false");
        std::env::remove_var("LOCALHOST_BYPASS");
        assert_eq!(require_user(&anon_req(&auth, Some("10.0.0.5"))).unwrap(), "");
        assert_eq!(require_user(&anon_req(&auth, None)).unwrap(), "");
        // A resolved user still wins (returned before the disabled check).
        let mut configured = anon_req(&auth, Some("10.0.0.5"));
        configured.current_user = Some("alice".to_string());
        assert_eq!(require_user(&configured).unwrap(), "alice");

        // --- 2. LOCALHOST_BYPASS=true + loopback: "" even when configured. ---
        std::env::set_var("AUTH_ENABLED", "true");
        std::env::set_var("LOCALHOST_BYPASS", "true");
        assert_eq!(require_user(&anon_req(&auth, Some("127.0.0.1"))).unwrap(), "");
        assert_eq!(require_user(&anon_req(&auth, Some("::1"))).unwrap(), "");
        assert_eq!(require_user(&anon_req(&auth, Some("localhost"))).unwrap(), "");
        // LOCALHOST_BYPASS does NOT help a non-loopback caller: it falls through
        // to the is_configured / first-run checks (only meaningful when the test
        // data dir is genuinely unconfigured).
        if auth.is_configured() {
            assert!(require_user(&anon_req(&auth, Some("10.0.0.5"))).is_err());
        }

        // --- 3. Plain modes (auth on, no bypass): loopback first-run "" vs 401. ---
        std::env::set_var("AUTH_ENABLED", "true");
        std::env::remove_var("LOCALHOST_BYPASS");
        if !auth.is_configured() {
            // Unconfigured first-run: loopback callers get "".
            assert_eq!(require_user(&anon_req(&auth, Some("127.0.0.1"))).unwrap(), "");
            // Non-loopback callers are rejected: HTTPException(401, ...).
            let e = require_user(&anon_req(&auth, Some("10.0.0.5"))).unwrap_err();
            assert!(matches!(e, PyError::Other(ref m) if m == "Not authenticated"));
        } else {
            // Configured: even loopback callers without a resolved user 401.
            assert!(require_user(&anon_req(&auth, Some("127.0.0.1"))).is_err());
        }

        restore("AUTH_ENABLED", &prev_auth);
        restore("LOCALHOST_BYPASS", &prev_bypass);
    }
}
